core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java

```diff
@@ -0,0 +1,276 @@
+                    continue;
+                }
+
+                buf[bufIdx++] = (char) (((byte) b << 6) ^ (byte) b2 ^
+                                        ((byte) 0xC0 << 6) ^ (byte) 0x80);
+                continue;
+            }
+
+            // 3-byte UTF-8
+            if ((b >>> 4) == 0b1110) {
+                if (undecodedChars < 9 || s.charAt(i) != '%' || s.charAt(i + 3) != '%') {
+                    buf[bufIdx++] = UNKNOWN_CHAR;
+                    i += 6;
+                    continue;
+                }
+
+                final int b2 = decodeHexByte(s.charAt(i + 1), s.charAt(i + 2));
+                final int b3 = decodeHexByte(s.charAt(i + 4), s.charAt(i + 5));
+                i += 6;
+
+                if (b2 < 0 || b3 < 0 ||
+                    (b == 0xe0 && (b2 & 0xe0) == 0x80) || !isContinuation(b2) || !isContinuation(b3)) {
+                    buf[bufIdx++] = UNKNOWN_CHAR;
+                    continue;
+                }
+
+                final char decoded = (char) (((byte) b << 12) ^ ((byte) b2 << 6) ^ (byte) b3 ^
+                                             ((byte) 0xE0 << 12) ^ ((byte) 0x80 << 6) ^ (byte) 0x80);
+                buf[bufIdx++] = !Character.isSurrogate(decoded) ? decoded : UNKNOWN_CHAR;
+                continue;
+            }
+
+            // 4-byte UTF-8
+            if ((b >>> 3) == 0b11110) {
+                if (undecodedChars < 12 ||
+                    s.charAt(i) != '%' || s.charAt(i + 3) != '%' || s.charAt(i + 6) != '%') {
+                    buf[bufIdx++] = UNKNOWN_CHAR;
+                    i += 9;
+                    continue;
+                }
+
+                final int b2 = decodeHexByte(s.charAt(i + 1), s.charAt(i + 2));
+                final int b3 = decodeHexByte(s.charAt(i + 4), s.charAt(i + 5));
+                final int b4 = decodeHexByte(s.charAt(i + 7), s.charAt(i + 8));
+                i += 9;
+
+                if (b2 < 0 || b3 < 0 || b4 < 0 ||
+                    !isContinuation(b2) || !isContinuation(b3) || !isContinuation(b4)) {
+                    buf[bufIdx++] = UNKNOWN_CHAR;
+                    continue;
+                }
+
+                final int codepoint =
+                        ((byte) b << 18) ^ ((byte) b2 << 12) ^ ((byte) b3 << 6) ^ (byte) b4 ^
+                        ((byte) 0xF0 << 18) ^ ((byte) 0x80 << 12) ^ ((byte) 0x80 << 6) ^ (byte) 0x80;
+                buf[bufIdx++] = Character.highSurrogate(codepoint);
+                buf[bufIdx++] = Character.lowSurrogate(codepoint);
+                continue;
+            }
+
+            buf[bufIdx++] = UNKNOWN_CHAR;
+        }
+
+        return new String(buf, 0, bufIdx);
+    }
+
+    private static int decodeHexByte(char c1, char c2) {
+        final int hi = decodeHexNibble(c1);
+        final int lo = decodeHexNibble(c2);
+        return (hi << 4) | lo;
+    }
+
+    private static int decodeHexNibble(char c) {
+        return c < OCTETS_TO_HEX.length ? OCTETS_TO_HEX[c] : -1;
```

Comment:
https://github.com/netty/netty/pull/9896 is coming too :)

We definitely should add a check vs 0 too to try to eliminate JVM's bounds check. I suspect converting to `int` first could be important since the value in the comparisons is automatically converted to `int` and it wouldn't surprise me if `(int) c` for the conditional and array indexing are treated as different values.

Reply:
Optimized